catalog/plugin/catalog-plugin-configuration-mapping/src/main/java/org/codice/ddf/catalog/plugin/configuration/mapping/ConfigurationMappingPlugin.java

```diff
@@ -60,35 +67,77 @@ public DeleteRequest process(DeleteRequest input)
         return input;
     }
 
-    public void setExpansionService(Expansion expansionService) {
-        this.expansionService = expansionService;
+    private void expandAttributes(List<Metacard> metacards) {
+
+        AttributeRuleSet attributeRuleSet = new AttributeRuleSet(expansionService);
+
+        for (Metacard metacard : metacards) {
+            attributeRuleSet.expandMetacard(metacard);
+        }
     }
 
-    private void expandAttributes(List<Metacard> metacards) {
-        for (Map.Entry<String, List<String[]>> expansion : expansionService.getExpansionMap()
-                .entrySet()) {
-            String attribute = expansion.getKey();
-            Map<String, String> expansionMap = toMap(expansion.getValue());
-
-            for (Metacard metacard : metacards) {
-                Attribute metacardAttribute = metacard.getAttribute(attribute);
-
-                if (metacardAttribute != null
-                        && expansionMap.containsKey(metacardAttribute.getValue())) {
-                    Attribute attributeImpl = new AttributeImpl(attribute,
-                            expansionMap.get(metacardAttribute.getValue()));
-                    metacard.setAttribute(attributeImpl);
+    private class AttributeRuleSet {
+
+        private Map<String, Map<String, String>> attributeRuleSet;
+
+        public AttributeRuleSet(Expansion expansion) {
+
+            attributeRuleSet = new HashMap<>();
+
+            for (Map.Entry<String, List<String[]>> attributeRules : expansion.getExpansionMap()
+                    .entrySet()) {
+                Map<String, String> attributeMap = toMap(attributeRules.getValue());
+
+                attributeRuleSet.put(attributeRules.getKey(), attributeMap);
+            }
+        }
+
+        public void expandMetacard(Metacard metacard) {
+
+            Set<String> attributeSet = metacard.getMetacardType()
+                    .getAttributeDescriptors()
+                    .stream()
+                    .map(des -> des.getName())
+                    .collect(Collectors.toSet());
+
+            Set<String> commonAttributes = new HashSet<>(CollectionUtils.intersection(attributeSet,
+                    attributeRuleSet.keySet()));
+
+            for (String attributeName : commonAttributes) {
+                Attribute metacardAttribute = metacard.getAttribute(attributeName);
+
+                if (metacardAttribute != null) {
+
+                    // expandTo holds the value that the ruleset
+                    String expandTo = getCorrespondingValue(attributeName,
+                            metacardAttribute.getValue());
```

Comment:
Unless I'm missing something, I think we need to loop through all the possible values in a metacard attribute since it can be a collection.

Say you want to expand `contact.name b -> bar`.
And the metacard attribute for `contact.name` contains a collection `a, b, c`. 

It will only check if `a` is in the `attributeRuleSet`, not expanding `b` to `bar`. This is per the `Attribute` interface javadoc, which states for `getValue` that `this method must return the first, if any, value`. If we instead use `getValues`, we can `getCorrespondingValue` on each of those to make sure we check everything and reach `b`. There should probably also be a test for this case.